### PR TITLE
Project search functionality in project resource page

### PR DESCRIPTION
**Pull request by Priyanka-Kaintura**
Targets #2960

### Description
Added the search functionality to the Project Resource page

### Checklist:

- [x] I have performed a self-review of my own code.

### CODE REVIEW

File: Modules/Project/Http/Controllers/ProjectController.php [hunk 1]
```diff
@@ -148,7 +148,7 @@ public function projectFTEExport(Request $request)
 
     public function projectResource()
     {
-        $resourceData = $this->service->getProjectsWithTeamMemberRequirementData();
+        $resourceData = $this->service->getProjectsWithTeamMemberRequirementData(request()->input('name'));
```
Comment (Ayush8923) [inline Modules/Project/Http/Controllers/ProjectController.php:151]:
```suggestion
        $resourceData = $this->service->getProjectsWithTeamMemberRequirementData(request()->all());
```

File: Modules/Project/Resources/views/resource-requirement.blade.php [hunk 2]
```diff
@@ -25,7 +34,7 @@
                 </thead>
                 <tbody>
                     @can('projects.view')
-                        @foreach ($resourceData['data'] as $clientName =>$clientProjectsData)
+                        @forelse ($resourceData['data'] as $clientName =>$clientProjectsData)
```
Comment (Ayush8923) [inline Modules/Project/Resources/views/resource-requirement.blade.php:37]: Why did you change `foreach` to `forelse` here?

File: Modules/Project/Services/ProjectService.php [hunk 3]
```diff
@@ -462,13 +462,17 @@ private function formatProjectFTEFOrExportAll($employees)
         return $teamMembers;
     }
 
-    public function getProjectsWithTeamMemberRequirementData()
+    public function getProjectsWithTeamMemberRequirementData($input)
```
Comment (Ayush8923) [inline Modules/Project/Services/ProjectService.php:465]: Please update the all places.
```suggestion
    public function getProjectsWithTeamMemberRequirementData($request)
```

File: Modules/Project/Services/ProjectService.php [hunk 4]
```diff
@@ -462,13 +462,17 @@ private function formatProjectFTEFOrExportAll($employees)
...
     {
-        $projectsWithTeamMemberRequirement = Project::with('client')->status('active')
+        $query = Project::with('client')->status('active')
```
Comment (Ayush8923) [inline Modules/Project/Services/ProjectService.php:467]: Please update the all places. Please use the correct naming convention.
```suggestion
        $projects = Project::with('client')->status('active')
```

**Review by Ayush8923 — CHANGES_REQUESTED** [review]
Left some comments.

File: Modules/Project/Resources/views/resource-requirement.blade.php (hunk 2 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Resources/views/resource-requirement.blade.php:37]: Because sir, the result of the search may even be null.

File: Modules/Project/Http/Controllers/ProjectController.php (hunk 1 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Http/Controllers/ProjectController.php:151]: Okay sir, I will do that.

File: Modules/Project/Services/ProjectService.php (hunk 3 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Services/ProjectService.php:465]: Okay sir.

File: Modules/Project/Services/ProjectService.php (hunk 4 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Services/ProjectService.php:467]: SIr, I didn't name it '$projects' because the resultant is not projects data, rather the query is aggregated further and then after applying the get() function we get the the collection as named in the code. So do I still need to rename the variable?

File: Modules/Project/Resources/views/resource-requirement.blade.php [hunk 5]
```diff
@@ -39,7 +48,15 @@
                             <tr>
                                 <td class="w-33p">
                                     <div class="pl-1 pl-xl-2">
-                                        {{ $projectName }}
+                                        @php
+                                            $project = $projectData['object'];
+                                            $team_member_ids = $project->getTeamMembers->pluck('team_member_id')->toArray();
+                                        @endphp
```
Comment (Ayush8923) [inline Modules/Project/Resources/views/resource-requirement.blade.php:54]: Please you can do all these logic operations in the service file. Then here you can use the variable directly.

File: Modules/Project/Services/ProjectService.php (hunk 4 above)
Comment (Ayush8923) [reply Modules/Project/Services/ProjectService.php:467]: Can we write like this? Please test locally once and also please test all the edge cases. and check this query work as expected.
Please let me know, it works or not.

```
$projectsWithTeamMemberRequirement = Project::query()
    ->with('client')
    ->status('active')
    ->withCount('getTeamMembers as team_member_count')
    ->withSum('resourceRequirement as team_member_needed', 'total_requirement')
    ->havingRaw('team_member_needed - team_member_count')
    ->when($request, function ($query, $request) {
        return $query->where('name', $request);
    })
    ->get();
```

File: Modules/Project/Services/ProjectService.php [hunk 6]
```diff
@@ -462,13 +462,17 @@ private function formatProjectFTEFOrExportAll($employees)
...
             ->withCount('getTeamMembers as team_member_count')
             ->withSum('resourceRequirement as team_member_needed', 'total_requirement')
-            ->havingRaw('team_member_needed - team_member_count')
-            ->get();
+            ->havingRaw('team_member_needed - team_member_count');
+
+        if ($request != null) {
+            $query->where('name', $request);
+        }
+        $projectsWithTeamMemberRequirement = $query->get();
```
Comment (Ayush8923) [inline Modules/Project/Services/ProjectService.php:475]: If you are using the above-suggested code, then please remove this code.

and also one more suggestion in this code. You can write like this if the condition. `empty` is a PHP function that checks whether a variable is empty or not.
```suggestion
        if ((!empty($request->name)) {
            $query->where('name', $request);
        }
```

**Review by Ayush8923 — CHANGES_REQUESTED** [review]
Left some suggestions @Priyanka-Kaintura.

File: Modules/Project/Services/ProjectService.php (hunk 6 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Services/ProjectService.php:475]: Yes Sir, I'll keep that in mind.

File: Modules/Project/Resources/views/resource-requirement.blade.php (hunk 5 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Resources/views/resource-requirement.blade.php:54]: sir, I don't know why these changes are visible here because these are the changes from previous pr which is now merged and closed. Please tell me if I still need to make these changes 
PR: https://github.com/ColoredCow/portal/pull/2927

File: Modules/Project/Services/ProjectService.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (Priyanka-Kaintura) [reply Modules/Project/Services/ProjectService.php:475]: Sir, it gives the result as expected and works fine. Thank You.

File: Modules/Project/Services/ProjectService.php (hunk 4 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Services/ProjectService.php:467]: Sir, it gives the result as expected and works fine. Thank You!

File: Modules/Project/Resources/views/resource-requirement.blade.php (hunk 5 above)
Comment (GauravGusain98) [reply Modules/Project/Resources/views/resource-requirement.blade.php:54]: Just take the pull from master branch and push it to this branch

File: Modules/Project/Services/ProjectService.php [hunk 7]
```diff
@@ -462,13 +462,18 @@ private function formatProjectFTEFOrExportAll($employees)
...
-            ->get();
+        $projectsWithTeamMemberRequirement = Project::query()
+        ->with('client')
+        ->status('active')
+        ->withCount('getTeamMembers as team_member_count')
+        ->withSum('resourceRequirement as team_member_needed', 'total_requirement')
+        ->havingRaw('team_member_needed - team_member_count')
+        ->when($request, function ($query, $request) {
+            return $query->where('name', $request['name']);
+        })
```
Comment (GauravGusain98) [inline Modules/Project/Services/ProjectService.php:475]: It should not be a strict check. Even a word or as substring is present in name, show that data. Also when we search an empty string then filter should not be applied. Similar to how we do in project page search

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]

File: Modules/Project/Resources/views/resource-requirement.blade.php (hunk 5 above)
Comment (Priyanka-Kaintura) [reply Modules/Project/Resources/views/resource-requirement.blade.php]: okay sir.

**Review by GauravGusain98 — APPROVED** [review]
LGTM
